src/txmempool.cpp

```diff
@@ -1556,6 +1586,85 @@ bool CTxMemPool::checkIncomingTxConflicts(const CTransaction& incomingTx) const
     return true;
 }
 
+bool CTxMemPool::certificateExists(const uint256& scId, int epochNumber) const
+{
+    LOCK(cs); // TODO: check this
+
+    if (epochNumber < 0)
+        return false;
+
+    for (const auto& cert_it: mapCertificate) {
+        const CScCertificate& cert = cert_it.second.GetCertificate();
+        if (cert.GetScId() == scId && cert.epochNumber == epochNumber)
+            return true;
+    }
+    return false;
+}
+
+bool CTxMemPool::checkReferencedHeight(const CScCertificate& incomingCert) const
+{
+    // If quality !=0, then for sure isNonCeasing() sidechain is false, so we do not need
+    // to perform any check.
+    if (incomingCert.quality != 0)
+        return true;
+
+    auto const& iheight_it = mapCumtreeHeight.find(incomingCert.endEpochCumScTxCommTreeRoot.GetLegacyHash());
+    // if we do not have this info, this must be a pre-v2 sc, hence ceasing
+    if (iheight_it == mapCumtreeHeight.end())
+        return true;
+
+    int icertHeight = iheight_it->second;
+    bool inserted = false;
+
+    LOCK(cs); // TODO: check this
+
+    const uint256& hash = incomingCert.GetHash();
+    const uint256& scId = incomingCert.GetScId();
+    auto const& sc_it = mapSidechains.find(scId);
+
+    if (sc_it == mapSidechains.end())
+        return true;
+
+    int64_t prevEpoch = -1;
+    int prevHeight = -1;
+
+    for (auto const& cert_it: sc_it->second.mBackwardCertificates)
+    {
+        // std::pair<int x, int y> is ordered as
+        // (x1,y1) > (x2,y2) if x1==x2 and y1 > y2
+
+        // retrieve referencedHeight
+        const CScCertificate& mapCert = mapCertificate.at(cert_it.second).GetCertificate();
+        if (mapCert.quality != 0)
+            return true; // isNonCeasing() == false
+
+        assert(mapCert.epochNumber > prevEpoch);
+
+        int mcertHeight = mapCumtreeHeight.at(mapCert.endEpochCumScTxCommTreeRoot.GetLegacyHash());
```

Comment:
Well, we are sure that this is a non ceasing sc, so there should be an entry, unless some referenced past blocks were invalidated. This can happen with commit 777a4efe60a61ad24940db040b81c68d5c33065d, so now we use a `find` here. Thanks.